linode/user/framework_datasource.go

```diff
@@ -0,0 +1,110 @@
+package user
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/hashicorp/terraform-plugin-framework/datasource"
+	"github.com/linode/linodego"
+	"github.com/linode/terraform-provider-linode/linode/helper"
+)
+
+func NewDataSource() datasource.DataSource {
+	return &DataSource{}
+}
+
+type DataSource struct {
+	client *linodego.Client
+}
+
+func (d *DataSource) Configure(
+	ctx context.Context,
+	req datasource.ConfigureRequest,
+	resp *datasource.ConfigureResponse,
+) {
+	// Prevent panic if the provider has not been configured.
+	if req.ProviderData == nil {
+		return
+	}
+
+	meta := helper.GetDataSourceMeta(req, resp)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	d.client = meta.Client
+}
+
+func (d *DataSource) Metadata(
+	ctx context.Context,
+	req datasource.MetadataRequest,
+	resp *datasource.MetadataResponse,
+) {
+	resp.TypeName = "linode_user"
+}
+
+func (d *DataSource) Schema(
+	ctx context.Context,
+	req datasource.SchemaRequest,
+	resp *datasource.SchemaResponse,
+) {
+	resp.Schema = frameworkDatasourceSchema
+}
+
+func (d *DataSource) Read(
+	ctx context.Context,
+	req datasource.ReadRequest,
+	resp *datasource.ReadResponse,
+) {
+	client := d.client
+
+	var data DataSourceModel
+
+	resp.Diagnostics.Append(req.Config.Get(ctx, &data)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	users, err := client.ListUsers(ctx, nil)
+	if err != nil {
+		resp.Diagnostics.AddError(
+			"Error listing Users: %s", err.Error(),
+		)
+		return
+	}
+
+	var user linodego.User
+
+	for _, testUser := range users {
+		if testUser.Username == data.Username.ValueString() {
+			user = testUser
+			break
+		}
+	}
+
+	if user.Username != "" {
+		resp.Diagnostics.Append(data.parseUser(ctx, &user)...)
+
+		if user.Restricted {
+			grants, err := client.GetUserGrants(ctx, data.Username.ValueString())
+			if err != nil {
+				resp.Diagnostics.AddError(
+					fmt.Sprintf("Failed to get User Grants (%s): ", data.Username.ValueString()), err.Error(),
+				)
+				return
+			}
+
+			resp.Diagnostics.Append(data.parseUserGrants(ctx, grants)...)
+		}
+
+		if resp.Diagnostics.HasError() {
+			return
+		}
+		resp.Diagnostics.Append(resp.State.Set(ctx, &data)...)
+	} else {
+		resp.Diagnostics.AddError(
+			fmt.Sprintf("Linode User with username %s was not found", data.Username.ValueString()), "",
+		)
+		return
+	}
```

Comment:
We can tidy this logic up a tiny bit

```suggestion
	if user.Username == "" {
		resp.Diagnostics.AddError(
			fmt.Sprintf("Linode User with username %s was not found", data.Username.ValueString()), "",
		)
		return
	}

	resp.Diagnostics.Append(data.parseUser(ctx, &user)...)

	if user.Restricted {
		grants, err := client.GetUserGrants(ctx, data.Username.ValueString())
		if err != nil {
			resp.Diagnostics.AddError(
				fmt.Sprintf("Failed to get User Grants (%s): ", data.Username.ValueString()), err.Error(),
			)
			return
		}

		resp.Diagnostics.Append(data.parseUserGrants(ctx, grants)...)
	}

	if resp.Diagnostics.HasError() {
		return
	}
	resp.Diagnostics.Append(resp.State.Set(ctx, &data)...)
```